### ISSUE TITLE
hard wrap prose at (parts of) sentences

### ISSUE THREAD

**Issue by maxheld83** [tidyverse/style#162]
To meet the 80 character line limit *where possible*, I would suggest to hard wrap prose, such as inside code comments or markdown at (in decreasing order of desirability):

1. Periods `.`
2. Colons `:`
3. Semicolon `;`
3. Commas `,`

If it is not possible, I would suggest allowing overflowing the 80 character limit.

This would be part of the norms of adopting a markdown style guide as per #161.
However, many popular markdown style guides are either silent on the issue, or include several options -- so there's a decision to be made.

My suggestion would be:

```md
<!--Good-->
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, 
sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, 
sed diam voluptua. 
At vero eos et accusam et justo duo dolores et ea rebum. 
```
(hard wrapping at above punctuation, otherwise overflowing 80 chars)

```md
<!--Also Good-->
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. 
At vero eos et accusam et justo duo dolores et ea rebum. 
```
(hard wrapping only at periods, otherwise overflowing 80 chars)

```md
<!--Bad-->
Lorem ipsum dolor sit amet, consetetur sadipscing elitr, 
sed diam nonumy eirmod tempor invidunt ut labore et dolore 
magna aliquyam erat, sed diam voluptua. At vero eos et 
accusam et justo duo dolores et ea rebum. 
```
(hard wrapping at 80 chars with no regard to punctuation)

---

#### Rationale

1. Style guides always attempt to visualise the syntax of the language in question, such as when using vertical alignment around control structures.
    The syntax of natural language is structured by punctuation.
2. Hard wrapping at 80 characters without regard to punctuation can cause confusing `git diff`s.
     For example, some change in line 13 can require require changes in the following lines because due to reflowing the hard breaks.
     In these (frequent) cases, the `git diff`s muddy the intent of the committer.
3. Guaranteed-at-any-cost hard wrapping at 80 characters was necessary for two reasons, which now/here apply to a lesser extend:
     1. Soft wrap wasn't always available. 
         It is now frequently available (editors, IDE, even some `git difftool`s), though admittedly still not everywhere (knitr, terminal).
     2. The overall horizontal organisation of code would be broken by soft wraps.
          This is less of a concern for (markdown) prose, which has little horizontal organisation to begin with.
            Admittedly, there is *some* horizontal organisation in markdown (nested lists) which might look worse when soft wrapped.

These arguments are shared by those (few) markdown style guides which cover the topic at all:

[Arctic Ice Studio](https://arcticicestudio.github.io/styleguide-markdown/rules/whitespace.html#maximum-line-length):

> Therefore, this guide advices to avoid using a character limit per line for flowing text, but try to use a maximum line length of 120 characters (including whitespaces) for all other document elements.

https://github.com/cirosantilli/markdown-style-guide/ (used by GitLab internally, apparently):

> Try to keep lines under 80 characters by breaking large paragraphs logically at points such as:
> - sentences: after a period ., question ? or exclamation mark !
> - clauses: after words like and, which, if ... then, commas ,
> large phrases
>
> It is acceptable to have a line longer than 80 characters, but keep in mind that long sentences are less readable and look worse in tools such as git diff.
>
> Set your editor to wrap lines visually for Markdown in case a large line is present.

I can also think of a **downside**: Hard wrapping at punctuation may conflict with the markdown design principle to be *readable-as-is*.

---

#### Existing Practice in the Tidyverse/r-lib

... is somewhat varied:

Overflow:

https://github.com/r-lib/roxygen2/blob/b3053b694399416543af877837efb29e6ec34d1f/vignettes/rd.Rmd#L183

Hard wrap (to maintain markdown syntax horizontal organisation):

https://github.com/r-lib/roxygen2/blob/b3053b694399416543af877837efb29e6ec34d1f/vignettes/rd.Rmd#L196-L200

Overflow (even when breaking markdown syntax):

https://github.com/r-lib/roxygen2/blob/b3053b694399416543af877837efb29e6ec34d1f/vignettes/rd.Rmd#L406-L408

**Comment by hadley** [comment]
Yeah, now I'd say just use the sentence wrapping provided by the IDE.
